src/Adapter/Product/Update/ProductShopUpdater.php

```diff
@@ -0,0 +1,306 @@
+
+        /** @var Product $sourceProduct */
+        $sourceProduct = $this->productRepository->get($productId, $sourceShopId);
+
+        // @todo: for now only fields from product_shop table ar handed, we still need to handle multilang, stock_available, and probably other things (in another PR)
+        $fields = [];
+        // The fields are fetched separately for more clarity, and it could also allow to configure which parts are copied
+        // (e.g copy only prices but not stock)
+        $fields = $fields + $this->getAssociationFields($sourceProduct);
+        $fields = $fields + $this->getPricesFields($sourceProduct);
+        $fields = $fields + $this->getStockFields($sourceProduct);
+        $fields = $fields + $this->getSEOFields($sourceProduct);
+        $fields = $fields + $this->getOptionsFields($sourceProduct);
+        $fields = $fields + $this->getCustomizationFields($sourceProduct);
+        $fields = $fields + $this->getDateFields($sourceProduct);
+
+        if ($this->productRepository->isAssociatedToShop($productId, $targetShopId)) {
+            $qb = $this->getUpdateQueryBuilder($fields, $productId, $targetShopId);
+        } else {
+            $qb = $this->getInsertQueryBuilder($fields, $productId, $targetShopId);
+        }
+
+        foreach ($fields as $column => $value) {
+            $qb->setParameter($column, $value);
+        }
+        $qb->execute();
+    }
+
+    /**
+     * @param array $fields
+     * @param ProductId $productId
+     * @param ShopId $targetShopId
+     *
+     * @return QueryBuilder
+     */
+    private function getInsertQueryBuilder(
+        array $fields,
+        ProductId $productId,
+        ShopId $targetShopId
+    ): QueryBuilder {
+        $fields['id_product'] = $productId->getValue();
+        $fields['id_shop'] = $targetShopId->getValue();
+
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->insert($this->dbPrefix . 'product_shop')
+            ->values($this->formatQueryValues($fields))
+            ->setParameter('id_product', $productId->getValue())
+            ->setParameter('id_shop', $targetShopId->getValue())
+        ;
+
+        return $qb;
+    }
+
+    /**
+     * @param array $fields
+     * @param ProductId $productId
+     * @param ShopId $targetShopId
+     *
+     * @return QueryBuilder
+     */
+    private function getUpdateQueryBuilder(
+        array $fields,
+        ProductId $productId,
+        ShopId $targetShopId
+    ): QueryBuilder {
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->update($this->dbPrefix . 'product_shop')
+            ->where('id_shop = :id_shop')
+            ->where('id_product = :id_product')
+            ->setParameter('id_shop', $targetShopId->getValue())
+            ->setParameter('id_product', $productId->getValue())
+        ;
+
+        $queryValues = $this->formatQueryValues($fields);
+        foreach ($queryValues as $column => $value) {
```

Comment:
So we fully avoid the use of ObjectModel here? That is a bold move 😄

Reply:
I'm not against it, I just hope it'll work without issues 👀

Reply:
@jolelievre - this is a big PR, so forgive me if that's somewhere in the code, but does it mean that dynamic hooks from the ObjectModel don't work anymore?

Reply:
Hi @kpodemski by dynamic hooks do you mean the ones triggered by the ObjectModel in save/add/update methods?

If you refer to these then they are still triggered when you edit the form don't worry, but indeed not in this peculiar case. However this feature didn't exist in the previous page, the idea here is that if you have a product assigned to a single shop (and not all the shops) you had no way to assign it to another one.

Now we will add a new feature that allows to assign the product to another by copying the data of a source shop. This is the purpose of this service, such feature didn't exist previously as far as I know.

I inspired most of it from the `ShopController` and how it handles the duplication of the products (and other entities) when you create a new shop, which relies mostly on this method:

https://github.com/PrestaShop/PrestaShop/blob/ca46aade5238ce31de77223217951180ce0cce2d/classes/shop/Shop.php#L1179

As you can see it uses direct SQL queries and does not rely on ObjectModels in this case (as it would cost too much performance I think with too many objects being created). So the ObjectModel related hooks are not lost here, because the feature is new and even in copy shop data they were not called.

But correct me if I'm wonrg on this maybe I missed some things regarding the product's hooks. Just by answering you in this comment I realized that I actually could/should integrate a hook in this service. Either a new one or maybe the one that is already present in the Shop method:

https://github.com/PrestaShop/PrestaShop/blob/ca46aade5238ce31de77223217951180ce0cce2d/classes/shop/Shop.php#L1271

The problem is that this hook is generic to the shop copy, not the product only so I guess a new hook would be better.

What do you think of all this?

Reply:
Legacy behavior is that if you save the product, even for the new store, the `$object->update` is being executed.

My experience is that most of the developers don't use `actionShopDataDuplication`. It's barely available in the modules I've seen. Devs rely on the `actionObjectProductAdd`, `actionObjectProductUpdate`, etc.

Unfortunately, I think that if we are copying a product from one store to another, we need to fire the dynamic hooks from the `ObjectModel`, because they are related to many modules such as featured products, etc.

IMHO the most important thing is that cache needs to be cleared, but of course, we can't just clear the entire cache of the store because it's precarious. 

I need to investigate this further, @Matt75 who works in the extension team, might be helpful on this one.

Reply:
@kpodemski indeed that makes sense, even though this feature is new I understand that it will be simpler for modules to integrate with it we keep on relying on the dynamic hooks

Good news is I was able to refacto this method quite quickly, by relying on the existing methods that were introduced in this PR, the copy method basically does:
- get product with shop constraint on source shop
- update product with shop constraint on target shop

For now it seems to work fine, at least all the behat tests regarding prices are still green, we'll probably need to look into more details for other scopes (by adding relevant behat tests), and especially regarding the associations but it's for another PR.

So now when a product is copied into another one the hooks `actionObjectUpdateBefore` (and after, and for product) will be triggered, even when the product was not previously associated to the shop.

Do you think it's ok this way, or should I make it so that the create hook is triggered instead? I think the update one is more suited but I prefer to be sure.